### ISSUE TITLE
SQL->whereNotIn() broken for empty arrays

### ISSUE THREAD

**Issue by mtschirs** [vanilla/vanilla#2969]
`Gdn_SqlDriver::whereNotIn('field', array())` produces the following SQL where clause:
`where field not in (null)`

Unfortunately, this condition always evaluates to false. This is contrary to what a developer might expect when writing e.g. `SQL->whereNotIn('d.DiscussionID', array());`. 

This is due to `Gdn_SqlDriver::_whereIn('field', array(), 'not in')` handling the special case of empty arrays with '(null)' which makes sense for the 'in' operator, but not the 'not in' operator.

See line 2045,Gdn_SqlDriver::_whereIn():

```
        if (count($In) > 0) {
            $InExpr = '('.implode(', ', $In).')';
        } else {
            $InExpr = '(null)';
        }
```

Solution: return without producing any SQL if the operator is 'not in' and count($In) = 0. Better handle this in a new method Gdn_SqlDriver::_whereNotIn().

**Comment by charrondev** [comment]
https://github.com/vanilla/vanilla/pull/7715 attempted to resolve this but the implementation broke backwards compatibility in a quite a few ways. There's probably some good additional context there if someone else picks it up.

**Comment by tburry** [comment]
Fixed by #9439.
